src/plugins/data/public/ui/dataset_navigator/dataset_navigator.tsx

```diff
@@ -485,83 +419,227 @@ export const DataSetNavigator = (props: DataSetNavigatorProps) => {
     </EuiPanel>
   );
 
-  const createIndexPatternsPanel = () => ({
-    id: 1,
-    title: indexPatternsLabel,
-    items: navigatorState.indexPatterns.map((indexPattern) => ({
-      name: indexPattern.title,
-      onClick: async () => await handleSelectedDataSet(indexPattern),
-    })),
-    content: <div>{navigatorState.indexPatterns.length === 0 && createLoadingSpinner()}</div>,
-  });
-
-  const createIndexesPanel = () => ({
-    id: 2,
-    title: indicesLabel,
-    items: [
-      ...navigatorState.dataSources.map((dataSource) => ({
-        name: dataSource.name,
-        panel: 3,
-        onClick: async () => await handleSelectedDataSource(dataSource),
-      })),
-    ],
-    content: <div>{navigatorState.isLoading && createLoadingSpinner()}</div>,
-  });
-
-  const createDataSourcesPanel = () => ({
-    id: 3,
-    title: navigatorState.currentDataSourceRef?.name ?? indicesLabel,
-    items: navigatorState.currentDataSourceRef?.indices?.map((object) => ({
-      name: object.title,
-      panel: 7,
-      onClick: async () => await handleSelectedObject(object),
-    })),
-    content: (
-      <div>
-        {navigatorState.isLoading && !navigatorState.currentDataSourceRef && createLoadingSpinner()}
-      </div>
-    ),
-  });
-
-  const createS3DataSourcesPanel = () => ({
-    id: 4,
-    title: (
-      <div>
-        {S3DataSourcesLabel}
-        {CatalogCacheManager.getExternalDataSourcesCache().status ===
-          CachedDataSourceStatus.Updated && createRefreshButton()}
-      </div>
-    ),
-    items: [
-      ...navigatorState.externalDataSources.map((dataSource) => ({
-        name: dataSource.name,
-        onClick: async () => await handleSelectExternalDataSource(dataSource),
-        panel: 5,
-      })),
-    ],
-    content: <div>{dataSourcesLoadStatus && createLoadingSpinner()}</div>,
-  });
-
-  const createDatabasesPanel = () => ({
-    id: 5,
-    title: selectedDataSetState?.dataSourceRef?.name
-      ? selectedDataSetState.dataSourceRef?.name
-      : 'Databases',
-    items: [
-      ...navigatorState.cachedDatabases.map((db) => ({
-        name: db.name,
-        onClick: async () => {
-          setSelectedDataSetState((prevState) => ({
-            ...prevState,
-            database: db,
-          }));
-          await handleSelectExternalDatabase(db);
-        },
-        panel: 6,
-      })),
+  const panels = useMemo(
```

Comment:
No need to change this now but the more I looked into the panels code, the more i'm convinced that we should only use one panel per nesting. And then we control the content of the panel based on what was selected in the previous nesting.

Reply:
+1, i agree, the state being more contained is safer.